relay-server/src/services/global_rate_limits.rs

```diff
@@ -0,0 +1,242 @@
+use relay_quotas::{
+    GlobalLimiter, GlobalRateLimiter, OwnedRedisQuota, RateLimitingError, RedisQuota,
+};
+use relay_redis::RedisPool;
+use relay_system::{
+    Addr, AsyncResponse, FromMessage, Interface, MessageResponse, Receiver, Sender, Service,
+};
+use std::sync::{Arc, Mutex};
+
+/// A request to check which quotas are rate limited by the global rate limiter.
+///
+/// This is typically sent to a [`GlobalRateLimitsService`] to determine which quotas
+/// should be rate limited based on the current usage.
+pub struct CheckRateLimited {
+    pub global_quotas: Vec<OwnedRedisQuota>,
+    pub quantity: usize,
+}
+
+/// Interface for global rate limiting operations.
+///
+/// This enum defines the messages that can be sent to a service implementing
+/// global rate limiting functionality.
+pub enum GlobalRateLimits {
+    /// Checks which quotas are rate limited by the global rate limiter.
+    CheckRateLimited(
+        CheckRateLimited,
+        Sender<Result<Vec<OwnedRedisQuota>, RateLimitingError>>,
+    ),
+}
+
+impl Interface for GlobalRateLimits {}
+
+impl FromMessage<CheckRateLimited> for GlobalRateLimits {
+    type Response = AsyncResponse<Result<Vec<OwnedRedisQuota>, RateLimitingError>>;
+
+    fn from_message(
+        message: CheckRateLimited,
+        sender: <Self::Response as MessageResponse>::Sender,
+    ) -> Self {
+        Self::CheckRateLimited(message, sender)
+    }
+}
+
+pub struct GlobalRateLimitsServiceHandle {
+    tx: Addr<GlobalRateLimits>,
+}
+
+impl GlobalLimiter for GlobalRateLimitsServiceHandle {
+    async fn check_global_rate_limits<'a>(
+        &self,
+        global_quotas: &'a [RedisQuota<'a>],
+        quantity: usize,
+    ) -> Result<Vec<&'a RedisQuota<'a>>, RateLimitingError> {
+        // We build the owned quotas to send over to the service.
+        let owned_global_quotas = global_quotas
+            .iter()
+            .map(|q| q.build_owned())
+            .collect::<Vec<_>>();
+
+        let rate_limited_owned_global_quotas = self
+            .tx
+            .send(CheckRateLimited {
+                global_quotas: owned_global_quotas,
+                quantity,
+            })
+            .await
+            .map_err(|_| RateLimitingError::UnreachableGlobalRateLimits)?;
+
+        // For each owned quota, we lookup the original quota and match it. In case of multiple
+        // duplicate quotas, the first one will be matched.
+        //
+        // This is done to make the `check_global_rate_limits` function depend only on references
+        // and hide the fact that we send owned data across threads.
+        //
+        // The operation is O(n^2) but we are assuming the number of quotas is bounded by a low number
+        // since now they are only used for metric buckets limiting.
+        let rate_limited_global_quotas =
```

Comment:
Not happy about this, but it felt like a good tradeoff between proper abstraction and performance. If we really want to avoid the backward mapping, we can directly return `OwnedRedisQuota`, as the caller side can support that, but it would tie the abstraction to the `Service` which requires the transfer of owned data.

A middle ground could be returning the indexes of the quotas that have been rate-limited, but this adds unnecessary complexity for relatively little benefit.